mr/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java

```diff
@@ -80,6 +81,10 @@ public void preCreateTable(org.apache.hadoop.hive.metastore.api.Table hmsTable)
         Preconditions.checkArgument(catalogProperties.getProperty(InputFormatConfig.PARTITION_SPEC) == null,
             "Iceberg table already created - can not use provided partition specification");
 
+        Schema hmsSchema = HiveSchemaUtil.schema(hmsTable.getSd().getCols());
+        Preconditions.checkArgument(HiveSchemaUtil.compatible(hmsSchema, icebergTable.schema()),
+            "Iceberg table already created - with different specification");
```

Comment:
What about rejecting the use of a schema if the table already exists in a different catalog? I think that would make more sense than allowing it but checking compatibility.

Reply:
The Hive API messes up with us. Here we always receive the Hive column definition which was generated based on the SerDe.getObjectInspector. The SerDe either uses the columns provided or the Iceberg table schema. We can try to find a way to massage through the API the information whether the source of the schema is an Iceberg table or provided in command line, but in the medium-long term I think we would like to decouple the Iceberg schema and the Hive schema anyway.

Just a few examples where decoupling could be useful:
- Iceberg UUID column - we might want to map to STRING / VARCHAR / CHAR or even BINARY
- Iceberg Time - we might want to find a way to at least read these columns to some of the Hive types
- We might want to create an Hive table which only contains a few column from the underlying Iceberg table

So instead of providing a throw-away temporary solution I decided to create the first implementation which checks the compatibility of the 2 schemas. The current implementation only allows exact matches, but it could be easily changed to allow differences if the ObjectInspectors are improved to handle the conversions.

Does this makes sense?
Thanks,
Peter

Reply:
Not sure I understand the usecases mentioned here. Why would the user need to specify a Hive schema when creating the Hive table even for the UUID type lets say? Shouldn't the deserializer be responsible for returning the Hive compatible schema?

E.g. for Avro tables, users do not need to specify a "string" column for enum types, the AvroSerDe maps it to string correctly.

Reply:
After discussing offline with @marton-bod we agreed that we should have stricter mapping. See: https://github.com/apache/iceberg/pull/1612#discussion_r526065544

There is still one remaining important use-case:
- The table is created before trying to access from Hive and the Catalog is HadoopTable/HadoopCatalog/CustomCatalog. When the user wants to create the Hive table it wants to use only a few columns from the table. Either because it has too many columns and wants to have a simplified table in Hive, or the Iceberg table contains columns which are not supported by Hive, like `time` or `fixed`.

Reply:
We have combinations of two sources of schemas when creating tables
1) `InputFormatConfig.TABLE_SCHEMA` (either provided by the user when creating table, or coming from an existing table)
2) Hive schema (provided by the user, or based on our previous comment seems like Hive will generate it based on ObjectInspector anyways)

There are a few problems I think of when it comes to Hive schemas when compared to Iceberg
1) Field names are always lowercase
2) Field are always nullable
3) Difference in supported types

So, I don't see a good reason why we should even look at the Hive schema when `InputFormatConfig.TABLE_SCHEMA` is present. Can we just disregard the Hive schema in such cases? I would prefer looking at Hive schema when thats the only schema information we have available. We won't need compatibility checks as there would be nothing to compare against.

For the use case to read a subset of columns from a non-Hive catalog table, can we just make the user specify the the column names they want to read? The user having to provide type information would be error-prone and especially tedious for nested column definitions. The user provided type information can also go stale overtime in reference to the source table.

Reply:
Good summary @shardulm94! Thanks!

> There are a few problems I think of when it comes to Hive schemas when compared to Iceberg
> 
> 1. Field names are always lowercase
> 2. Field are always nullable
> 3. Difference in supported types

And we should add different partitioning philosophy, but that will be something for later 😄 
Good point on the differences. Definitely we should create a PR handling all those differences.

> So, I don't see a good reason why we should even look at the Hive schema when `InputFormatConfig.TABLE_SCHEMA` is present. Can we just disregard the Hive schema in such cases? I would prefer looking at Hive schema when thats the only schema information we have available. We won't need compatibility checks as there would be nothing to compare against.
> 
> For the use case to read a subset of columns from a non-Hive catalog table, can we just make the user specify the the column names they want to read? The user having to provide type information would be error-prone and especially tedious for nested column definitions. The user provided type information can also go stale overtime in reference to the source table.

For the longer term:
1. I think we should stick to the SQL standard as much as possible. I think it would be an overkill to introduce new SQL language elements only for the case where the Iceberg table is already created and we are creating a new Hive table with fever columns.
2. If the above point stands, then we have a schema from the SQL and the schema from the table which we have to compare anyway.

For short term:
1. We can decide to stick to the Hive schema generated from the Iceberg schema if we accept that Hive recreates the SerDe quite often and we have to load the table every case. This is "only" a performance question.
2. The bigger problem is that there is no technically correct way to differentiate cases when the column specification is provided by the user or when it is generated based on the Iceberg schema. We can rely on the column comments ("from deserializer") or we can use undocumented feature to add a new serDeProperties, but I would no like to depend on these for features. This means that we either have to compare the Hive schema and the Iceberg schema, or we have to throw away the schema provided by Hive. This later solution could be a usability problem since we will hide the case when the user provided a schema and it is not the same as the schema generated from the Iceberg schema.

So my thought process was the following:
1. In the long term we need the schema comparison
2. In the short term it would be good to have a schema comparison to provide good usability
3. This is why I decided to implement the comparison.

The options I see:
1. Move forward with the change as it is 😄 
2. Remove comparison and if there is an Iceberg schema available for us, use it even if the user provided columns. We can remove the unsupported columns / struct fields / maps, so the Hive table could be created. This could raise some eyebrows when the table specification is provided in the SQL and the created table does not conform to that specification, but we can document this "feature" 😄 
3. I am open to other suggestions

Thanks,
Peter